ffi/src/winscard/scard.rs

```diff
@@ -40,34 +118,60 @@ pub extern "system" fn SCardReconnect(
     _dw_initialization: u32,
     _pdw_active_protocol: LpDword,
 ) -> ScardStatus {
-    todo!()
+    ErrorKind::UnsupportedFeature.into()
 }
 
 #[cfg_attr(windows, rename_symbol(to = "Rust_SCardDisconnect"))]
+#[instrument(ret)]
 #[no_mangle]
-pub extern "system" fn SCardDisconnect(_handle: ScardHandle, _dw_disposition: u32) -> ScardStatus {
-    todo!()
+pub unsafe extern "system" fn SCardDisconnect(handle: ScardHandle, _dw_disposition: u32) -> ScardStatus {
+    check_handle!(handle);
+
+    let scard = Box::from_raw(handle as *mut WinScardHandle);
+    if let Some(context) = (scard.context() as *mut WinScardContextHandle).as_mut() {
+        if context.remove_scard(handle) {
+            info!(?handle, "Successfully disconnected!");
+        } else {
+            warn!("ScardHandle does not belong to the specified context.")
+        }
+    }
+
+    ErrorKind::Success.into()
 }
 
 #[cfg_attr(windows, rename_symbol(to = "Rust_SCardBeginTransaction"))]
+#[instrument(ret)]
 #[no_mangle]
-pub extern "system" fn SCardBeginTransaction(_handle: ScardHandle) -> ScardStatus {
-    todo!()
+pub unsafe extern "system" fn SCardBeginTransaction(handle: ScardHandle) -> ScardStatus {
+    check_handle!(handle);
+    let scard = try_execute!(scard_handle_to_winscard(handle));
+
+    try_execute!(scard.begin_transaction());
+
+    ErrorKind::Success.into()
 }
 
 #[cfg_attr(windows, rename_symbol(to = "Rust_SCardEndTransaction"))]
+#[instrument(ret)]
 #[no_mangle]
-pub extern "system" fn SCardEndTransaction(_handle: ScardHandle, _dw_disposition: u32) -> ScardStatus {
-    todo!()
+pub unsafe extern "system" fn SCardEndTransaction(handle: ScardHandle, _dw_disposition: u32) -> ScardStatus {
+    check_handle!(handle);
+    let scard = try_execute!(scard_handle_to_winscard(handle));
+
+    try_execute!(scard.end_transaction());
+
+    ErrorKind::Success.into()
 }
```

Comment:
suggestion: Ideally, I would like unsafe blocks even inside "unsafe function": https://doc.rust-lang.org/rustc/lints/listing/allowed-by-default.html#unsafe-op-in-unsafe-fn
This will be warn-by-default or even deny-by-default in future editions: https://github.com/rust-lang/rust/issues/71668

note: The function should be unsafe only if the caller must unhold invariants himself. In this case, if there is an invariant on `ScardHandle`, then the function being marked unsafe is good. A typical example is when the caller must ensure a pointer is non-null. If `ScardHandle` is always valid by definition (e.g.: the user can’t craft an invalid `ScardHandle` and we always return a valid one), then the function is not unsafe to call per-se.

Reply:
Good point! I thought about it for a while and made the following changes:

1. I made `scard_handle_to_winscard`, `scard_context_to_winscard_context`, `copy_buff`, and other similar functions also `unsafe`. They all work with raw pointers provided by the user and we can't guarantee any safe behaviour.
2. I added the `#[deny(unsafe_op_in_unsafe_fn)]` lint but for the `winscard` module only. I think we can refactor the `sspi` module as well, but it's better to do it in a separate PR.

Reply:
Sounds good to me! Thank you!!